File: jinwoo/programmers-2/가장 큰 정사각형 찾기.java
class Solution
{
    public int solution(int[][] board) {
        int answer = 0;
        //주어진 배열에서 첫번째 행과 첫번째 열에 1이 있는지 체크하기 위한 변수
        //아래의 반복문에서는 첫번째행과 첫번째열의 값을 비교해나가며 배열의 값을 바꾸기 때문에 두번째행과 두번째열부터 값을 바꿔나가고 있음.
        //따라서 첫번째행과 첫번째열에 1X1 크기의 정사각형이 있는지 별도로 체크
        boolean one = false;

        //첫번째행에서 1이 있는지 체크
        for (int i = 0; i < board[0].length; i++) {
            if (board[0][i] == 1) one = true;
        }
        //첫번째열에서 1이 있는지 체크
        if (one == false) {
            for (int i = 0; i < board.length; i++) {
                if (board[i][0] == 1) one = true;
            }
        }
        for (int i = 1; i < board.length; i++) {
            for (int j = 1; j < board[0].length; j++) {
                if (board[i][j] != 0)
                    board[i][j] = Math.min(Math.min(board[i - 1][j], board[i][j - 1]), board[i - 1][j - 1]) + 1;
                answer = Math.max(board[i][j], answer);
            }
        }
        if (answer==0 && one==true) answer=1;
        return answer * answer;
    }
}

File: seungho/programmers-2/가장-큰-정사각형-찾기.ts
/*
1와 0로 채워진 표(board)가 있습니다. 표 1칸은 1 x 1 의 정사각형으로 이루어져 있습니다. 표에서 1로 이루어진 가장 큰 정사각형을 찾아 넓이를 return 하는 solution 함수를 완성해 주세요. (단, 정사각형이란 축에 평행한 정사각형을 말합니다.)

예를 들어

1	2	3	4
0	1	1	1
1	1	1	1
1	1	1	1
0	0	1	0
가 있다면 가장 큰 정사각형은

1	2	3	4
0	1	1	1
1	1	1	1
1	1	1	1
0	0	1	0
가 되며 넓이는 9가 되므로 9를 반환해 주면 됩니다.

제한사항
표(board)는 2차원 배열로 주어집니다.
표(board)의 행(row)의 크기 : 1,000 이하의 자연수
표(board)의 열(column)의 크기 : 1,000 이하의 자연수
표(board)의 값은 1또는 0으로만 이루어져 있습니다.
*/

const findLargestSquare = (board: number[][]): number => {
    const row: number = board.length;
    const col: number = board[0].length;
    let maxSide: number = 0;

    if (row === 1 || col === 1) {
        return Math.max(...board.flat());
    }

    for (let i = 1; i < row; i++) {
        for (let j = 1; j < col; j++) {
            if (board[i][j] === 1) {
                board[i][j] = Math.min(board[i-1][j], board[i][j-1], board[i-1][j-1]) + 1;
                maxSide = Math.max(maxSide, board[i][j]);
            }
        }
    }

    return maxSide ** 2;
}